<option> label attribute updates are not shown until focus is moved

krzysdz opened the pull request:
#### 06461c59e1c13e5569601f8defb7eadc46459813
<pre>
&lt;option&gt; label attribute updates are not shown until focus is moved
<a href="https://bugs.webkit.org/show_bug.cgi?id=258862">https://bugs.webkit.org/show_bug.cgi?id=258862</a>

Reviewed by Tim Nguyen.

HTMLOptionElement::label updates from JavaScript were not rendered on parent
&lt;select&gt; element, unless focus was moved to/from it. This change informs the
HTMLSelectElement about &lt;option&gt; child update whenever the label attribute
changes, which allows it to render the updated label.

* LayoutTests/TestExpectations:
* Source/WebCore/html/HTMLOptionElement.cpp:
(WebCore::HTMLOptionElement::attributeChanged):

Canonical link: <a href="https://commits.webkit.org/265942@main">https://commits.webkit.org/265942@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2a942e427a1fd66adb385cce86f22b16b1c6caa1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12260 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12604 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12908 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14005 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11821 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12290 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15023 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12621 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14497 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12424 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13210 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10382 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14425 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10495 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11123 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18236 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11576 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11287 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14469 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11780 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9718 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10998 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3035 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15330 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11634 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/fast/forms/select/selected-option-update-label.html:
Would be nice if this was a web platform test

krzysdz replied:
I'm sorry, but I don't know how to rewrite this test as a web platform test (aren't reftests web platform tests?). If you can give me a hint, I'll try to do it.

Ahmad-S792 replied:
@krzysdz - I think this would be helpful - https://web-platform-tests.org/index.html and looking into 'Writing Tests'. :-)

krzysdz replied:
I've checked the existing Web Platform Tests and it turns out that similar tests already exist:
- [`html/rendering/widgets/the-select-element/option-add-label.html`](https://github.com/WebKit/WebKit/blob/8b1a13b39bbaaf306c9d819c13b0811011be55f2/LayoutTests/imported/w3c/web-platform-tests/html/rendering/widgets/the-select-element/option-add-label.html)
- [`html/rendering/widgets/the-select-element/option-rm-label.html`](https://github.com/WebKit/WebKit/blob/8b1a13b39bbaaf306c9d819c13b0811011be55f2/LayoutTests/imported/w3c/web-platform-tests/html/rendering/widgets/the-select-element/option-rm-label.html)

I can't find them in the EWS results, but when running locally they pass:
![WPT test passed unexpectedly](https://github.com/WebKit/WebKit/assets/12915102/dcec6893-12bd-4dc8-a8cc-c0a63b1eb5e8)

Should I then just remove this test?

nt1m replied:
You probably just want to remove these two lines:
https://searchfox.org/wubkat/rev/ff57fb336bb501c2d8d318ccfe8124678439fccc/LayoutTests/TestExpectations#827,830

and yeah the new test you added is probably not needed

krzysdz committed:
<option> label attribute updates are not shown until focus is moved
https://bugs.webkit.org/show_bug.cgi?id=258862

Reviewed by Tim Nguyen.

HTMLOptionElement::label updates from JavaScript were not rendered on parent
<select> element, unless focus was moved to/from it. This change informs the
HTMLSelectElement about <option> child update whenever the label attribute
changes, which allows it to render the updated label.

* LayoutTests/TestExpectations:
* Source/WebCore/html/HTMLOptionElement.cpp:
(WebCore::HTMLOptionElement::attributeChanged):

Canonical link: https://commits.webkit.org/265942@main
